From 4212e237ba488747fb918224b12da05bed4dfd8e Mon Sep 17 00:00:00 2001
Message-Id: <4212e237ba488747fb918224b12da05bed4dfd8e.1423172493.git.feitong.yi@intel.com>
In-Reply-To: <18bb7324f916bffbc75e07ea0ffd3e6e68e7b6d8.1423172493.git.feitong.yi@intel.com>
References: <18bb7324f916bffbc75e07ea0ffd3e6e68e7b6d8.1423172493.git.feitong.yi@intel.com>
From: rtshiva <sivakumar.thulasimani@intel.com>
Date: Wed, 14 Jan 2015 08:41:17 +0530
Subject: [PATCH 17/29] MUST_REBASE [VPG]: adf/intel/vlv: Add the port
 structure for dsi encoder

This patch adds new file for dsi_encoder port which contains all
the operations related to dsi port programming.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

For: GMINL-5213
Change-Id: I78c9d9464c7a8d7f11edc25637154a6b054784f1
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Signed-off-by: Deepak M <m.deepak@intel.com>
---
 drivers/video/adf/intel/core/vlv/Makefile          |    4 +-
 drivers/video/adf/intel/core/vlv/vlv_dsi_port.c    |  468 ++++++++++++++++++++
 .../video/adf/intel/core/vlv/vlv_dsi_port_cmd.c    |  386 ++++++++++++++++
 .../adf/intel/include/core/vlv/vlv_dc_config.h     |    5 +
 .../adf/intel/include/core/vlv/vlv_dsi_port.h      |  100 +++++
 5 files changed, 962 insertions(+), 1 deletion(-)
 create mode 100644 drivers/video/adf/intel/core/vlv/vlv_dsi_port.c
 create mode 100644 drivers/video/adf/intel/core/vlv/vlv_dsi_port_cmd.c
 create mode 100644 drivers/video/adf/intel/include/core/vlv/vlv_dsi_port.h

diff --git a/drivers/video/adf/intel/core/vlv/Makefile b/drivers/video/adf/intel/core/vlv/Makefile
index a3e8f2a..6c4dd85 100644
--- a/drivers/video/adf/intel/core/vlv/Makefile
+++ b/drivers/video/adf/intel/core/vlv/Makefile
@@ -11,4 +11,6 @@ obj-y := vlv_dc_config.o \
 	vlv_pipe.o \
 	vlv_pll.o \
 	vlv_dsi_pll.o \
-	vlv_pm.o
+	vlv_pm.o \
+	vlv_dsi_port.o \
+	vlv_dsi_port_cmd.o
diff --git a/drivers/video/adf/intel/core/vlv/vlv_dsi_port.c b/drivers/video/adf/intel/core/vlv/vlv_dsi_port.c
new file mode 100644
index 0000000..e332035
--- /dev/null
+++ b/drivers/video/adf/intel/core/vlv/vlv_dsi_port.c
@@ -0,0 +1,468 @@
+/*
+ * Copyright (C) 2014, Intel Corporation.
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <intel_adf_device.h>
+#include <drm/drmP.h>
+#include <core/common/dsi/dsi_pipe.h>
+#include <core/vlv/vlv_dc_config.h>
+#include <core/vlv/vlv_dsi_port.h>
+
+
+u32 vlv_dsi_port_is_vid_mode(struct vlv_dsi_port *port)
+{
+	return port->operation_mode == DSI_DPI;
+}
+
+u32 vlv_dsi_port_is_cmd_mode(struct vlv_dsi_port *port)
+{
+	return port->operation_mode == DSI_DBI;
+}
+
+u32  vlv_dsi_port_wait_for_fifo_empty(struct vlv_dsi_port *port)
+{
+	u32 mask;
+
+	mask = LP_CTRL_FIFO_EMPTY | HS_CTRL_FIFO_EMPTY |
+			LP_DATA_FIFO_EMPTY | HS_DATA_FIFO_EMPTY;
+
+	/* Wait for FIFOs to become empty */
+	if (wait_for((REG_READ(port->fifo_stat_offset) & mask) == mask, 100))
+		pr_err("ADF: %s DSI FIFO not empty\n", __func__);
+	return 0;
+}
+
+/* return txclkesc cycles in terms of divider and duration in us */
+static u16 txclkesc(u32 divider, unsigned int us)
+{
+	switch (divider) {
+	case ESCAPE_CLOCK_DIVIDER_1:
+	default:
+		return 20 * us;
+	case ESCAPE_CLOCK_DIVIDER_2:
+		return 10 * us;
+	case ESCAPE_CLOCK_DIVIDER_4:
+		return 5 * us;
+	}
+}
+
+/* return pixels in terms of txbyteclkhs */
+static u16 txbyteclkhs(u16 pixels, int bpp, int lane_count,
+			u16 burst_mode_ratio)
+{
+	if (lane_count)
+		return DIV_ROUND_UP(DIV_ROUND_UP(pixels * bpp *
+				burst_mode_ratio, 8 * 100), lane_count);
+	else
+		return -EINVAL;
+}
+
+/* to be called before pll disable */
+bool vlv_dsi_port_clear_device_ready(struct vlv_dsi_port *port)
+{
+	u32 val;
+
+	REG_WRITE(port->dr_offset, DEVICE_READY | ULPS_STATE_ENTER);
+	usleep_range(2000, 2500);
+
+	REG_WRITE(port->dr_offset, DEVICE_READY | ULPS_STATE_EXIT);
+	usleep_range(2000, 2500);
+
+	REG_WRITE(port->dr_offset, DEVICE_READY | ULPS_STATE_ENTER);
+	usleep_range(2000, 2500);
+
+	val = REG_READ(port->offset);
+	REG_WRITE(port->offset, val & ~LP_OUTPUT_HOLD);
+	usleep_range(1000, 1500);
+
+	if (wait_for(((REG_READ(port->offset) & AFE_LATCHOUT)
+			== 0x00000), 30))
+		pr_err("DSI LP not going Low\n");
+	REG_WRITE(port->dr_offset, 0x00);
+	usleep_range(2000, 2500);
+
+	return true;
+}
+
+bool vlv_dsi_port_set_device_ready(struct vlv_dsi_port *port)
+{
+	u32 val;
+
+	/* ULPS */
+	REG_WRITE(port->dr_offset, DEVICE_READY | ULPS_STATE_ENTER);
+	usleep_range(1000, 1500);
+
+	val = REG_READ(port->offset);
+	REG_WRITE(port->offset, val | LP_OUTPUT_HOLD);
+	usleep_range(1000, 1500);
+
+	if (wait_for(((REG_READ(port->offset) & AFE_LATCHOUT)
+				== 0x00000), 30))
+		pr_err("ADF: %s DSI LP not going Low\n", __func__);
+
+	REG_WRITE(port->dr_offset, DEVICE_READY | ULPS_STATE_EXIT);
+	usleep_range(2000, 2500);
+	REG_WRITE(port->dr_offset, DEVICE_READY);
+	usleep_range(2000, 2500);
+
+	return true;
+}
+
+
+static void set_dsi_timings(struct vlv_dsi_port *port,
+		struct drm_mode_modeinfo *mode,
+		struct dsi_config *config)
+{
+	struct dsi_context *intel_dsi = &config->ctx;
+	unsigned int bpp = config->bpp;
+	unsigned int lane_count = intel_dsi->lane_count;
+	u16 hactive, hfp, hsync, hbp, vfp, vsync, vbp;
+
+	hactive = mode->hdisplay;
+	hfp = mode->hsync_start - mode->hdisplay;
+	hsync = mode->hsync_end - mode->hsync_start;
+	hbp = mode->htotal - mode->hsync_end;
+
+	vfp = mode->vsync_start - mode->vdisplay;
+	vsync = mode->vsync_end - mode->vsync_start;
+	vbp = mode->vtotal - mode->vsync_end;
+
+	/* horizontal values are in terms of high speed byte clock */
+	hactive = txbyteclkhs(hactive, bpp, lane_count,
+			intel_dsi->burst_mode_ratio);
+	hfp = txbyteclkhs(hfp, bpp, lane_count, intel_dsi->burst_mode_ratio);
+	hsync = txbyteclkhs(hsync, bpp, lane_count,
+			intel_dsi->burst_mode_ratio);
+	hbp = txbyteclkhs(hbp, bpp, lane_count, intel_dsi->burst_mode_ratio);
+
+	REG_WRITE(port->hactive_count_offset, hactive);
+	REG_WRITE(port->hfp_count_offset, hfp);
+
+	/*
+	 * meaningful for video mode non-burst sync
+	 * pulse mode only, can be zero
+	 * for non-burst sync events and burst modes
+	 */
+	REG_WRITE(port->hsync_count_offset, hsync);
+	REG_WRITE(port->hbp_count_offset, hbp);
+
+	/* vertical values are in terms of lines */
+	REG_WRITE(port->vfp_count_offset, vfp);
+	REG_WRITE(port->vsync_count_offset, vsync);
+	REG_WRITE(port->vbp_count_offset, vbp);
+
+}
+
+u32  vlv_dsi_port_prepare(struct vlv_dsi_port *port,
+		struct drm_mode_modeinfo *mode,
+		struct dsi_config *config)
+{
+	struct dsi_context *intel_dsi = &config->ctx;
+	unsigned int bpp = config->bpp;
+	u32 val, tmp;
+
+	REG_WRITE(port->dr_offset, 0x0);
+	usleep_range(2000, 2500);
+
+	/*
+	 * escape clock divider, 20MHz, shared for A and C. device ready must be
+	 * off when doing this! txclkesc?
+	 */
+	/* FIXEME: check if this is applicable only for pipe A or not  */
+	tmp = REG_READ(port->ctrl_offset);
+	tmp &= ~ESCAPE_CLOCK_DIVIDER_MASK;
+	REG_WRITE(port->ctrl_offset, tmp | ESCAPE_CLOCK_DIVIDER_1);
+
+	/* read request priority is per pipe */
+	tmp = REG_READ(port->ctrl_offset);
+	tmp &= ~READ_REQUEST_PRIORITY_MASK;
+	REG_WRITE(port->ctrl_offset, tmp | READ_REQUEST_PRIORITY_HIGH);
+
+	/* XXX: why here, why like this? handling in irq handler?! */
+	/* FIXME: move these to common */
+	REG_WRITE(MIPI_INTR_STAT(0), 0xffffffff);
+	REG_WRITE(MIPI_INTR_EN(0), 0xffffffff);
+
+	REG_WRITE(port->dphy_param_offset, intel_dsi->dphy_reg);
+
+	REG_WRITE(port->dpi_res_offset,
+		mode->vdisplay << VERTICAL_ADDRESS_SHIFT |
+		mode->hdisplay << HORIZONTAL_ADDRESS_SHIFT);
+
+	set_dsi_timings(port, mode, config);
+
+	val = intel_dsi->lane_count << DATA_LANES_PRG_REG_SHIFT;
+	if (vlv_dsi_port_is_cmd_mode(port)) {
+		val |= intel_dsi->channel << CMD_MODE_CHANNEL_NUMBER_SHIFT;
+		val |= CMD_MODE_DATA_WIDTH_8_BIT; /* XXX */
+	} else {
+		val |= intel_dsi->channel << VID_MODE_CHANNEL_NUMBER_SHIFT;
+
+		/* XXX: cross-check bpp vs. pixel format? */
+		val |= intel_dsi->pixel_format;
+	}
+
+	val |= VID_MODE_NOT_SUPPORTED;
+	REG_WRITE(port->func_prg_offset, val);
+
+	/*
+	 * timeouts for recovery. one frame IIUC. if counter expires, EOT and
+	 * stop state.
+	 */
+
+	/*
+	 * In burst mode, value greater than one DPI line Time in byte clock
+	 * (txbyteclkhs) To timeout this timer 1+ of the above said value is
+	 * recommended.
+	 *
+	 * In non-burst mode, Value greater than one DPI frame time in byte
+	 * clock(txbyteclkhs) To timeout this timer 1+ of the above said value
+	 * is recommended.
+	 *
+	 * In DBI only mode, value greater than one DBI frame time in byte
+	 * clock(txbyteclkhs) To timeout this timer 1+ of the above said value
+	 * is recommended.
+	 */
+
+	if (vlv_dsi_port_is_vid_mode(port) &&
+	intel_dsi->video_mode_format == VIDEO_MODE_BURST) {
+		REG_WRITE(port->hs_tx_timeout_offset,
+			txbyteclkhs(mode->htotal, bpp,
+				intel_dsi->lane_count,
+				intel_dsi->burst_mode_ratio) + 1);
+	} else {
+		REG_WRITE(port->hs_tx_timeout_offset,
+			txbyteclkhs(mode->vtotal *
+				mode->htotal,
+				bpp, intel_dsi->lane_count,
+				intel_dsi->burst_mode_ratio) + 1);
+	}
+
+	REG_WRITE(port->lp_rx_timeout_offset, intel_dsi->lp_rx_timeout);
+	REG_WRITE(port->ta_timeout_offset, intel_dsi->turn_arnd_val);
+	REG_WRITE(port->device_reset_timer_offset, intel_dsi->rst_timer_val);
+
+	/* dphy stuff */
+
+	/* in terms of low power clock */
+	REG_WRITE(port->init_count_offset, txclkesc(intel_dsi->escape_clk_div,
+						100));
+
+	val = 0;
+	if (intel_dsi->eotp_pkt == 0)
+		val |= EOT_DISABLE;
+
+	if (intel_dsi->clock_stop)
+		val |= CLOCKSTOP;
+
+	/* recovery disables */
+	REG_WRITE(port->eot_offset, val);
+
+	/* in terms of low power clock */
+	REG_WRITE(port->init_count_offset, intel_dsi->init_count);
+
+	/*
+	 * in terms of txbyteclkhs. actual high to low switch +
+	 * MIPI_STOP_STATE_STALL * MIPI_LP_BYTECLK.
+	 *
+	 * XXX: write MIPI_STOP_STATE_STALL?
+	 */
+	REG_WRITE(port->hl_switch_count_offset, intel_dsi->hs_to_lp_count);
+
+	/*
+	 * XXX: low power clock equivalence in terms of byte clock. the number
+	 * of byte clocks occupied in one low power clock. based on txbyteclkhs
+	 * and txclkesc. txclkesc time / txbyteclk time * (105 +
+	 * MIPI_STOP_STATE_STALL) / 105.???
+	 */
+	REG_WRITE(port->lp_byteclk_offset, intel_dsi->lp_byte_clk);
+
+	/*
+	 * the bw essential for transmitting 16 long packets containing 252
+	 * bytes meant for dcs write memory command is programmed in this
+	 * register in terms of byte clocks. based on dsi transfer rate and the
+	 * number of lanes configured the time taken to transmit 16 long packets
+	 * in a dsi stream varies.
+	 */
+	REG_WRITE(port->dbi_bw_offset, intel_dsi->bw_timer);
+
+	REG_WRITE(port->lane_switch_time_offset,
+		intel_dsi->clk_lp_to_hs_count << LP_HS_SSW_CNT_SHIFT |
+		intel_dsi->clk_hs_to_lp_count << HS_LP_PWR_SW_CNT_SHIFT);
+
+	if (vlv_dsi_port_is_vid_mode(port))
+		/*
+		 * Some panels might have resolution which is not a multiple of
+		 * 64 like 1366 x 768. Enable RANDOM resolution support for such
+		 * panels by default
+		 */
+		REG_WRITE(port->video_mode_offset,
+				intel_dsi->video_frmt_cfg_bits |
+				intel_dsi->video_mode_format |
+				IP_TG_CONFIG |
+				RANDOM_DPI_DISPLAY_RESOLUTION);
+
+	REG_WRITE(port->dr_offset, DEVICE_READY);
+	usleep_range(2000, 2500);
+
+	return 0;
+}
+
+u32 vlv_dsi_port_pre_enable(struct vlv_dsi_port *port,
+			struct drm_mode_modeinfo *mode,
+			struct dsi_config *config)
+{
+	u32 val;
+	struct dsi_context *intel_dsi = &config->ctx;
+	vlv_dsi_port_wait_for_fifo_empty(port);
+
+	REG_WRITE(port->dr_offset, 0);
+	usleep_range(2000, 2500);
+
+	val = REG_READ(port->func_prg_offset);
+	val |= intel_dsi->pixel_format;
+	REG_WRITE(port->func_prg_offset, val);
+
+	val = REG_READ(port->eot_offset);
+	if (intel_dsi->eotp_pkt == 0)
+		val |= EOT_DISABLE;
+
+	REG_WRITE(port->eot_offset, val);
+
+	REG_WRITE(port->dr_offset, DEVICE_READY);
+	usleep_range(2000, 2500);
+
+	if (vlv_dsi_port_is_cmd_mode(port))
+		REG_WRITE(port->max_ret_pkt_offset, 8 * 4);
+
+	return 0;
+}
+
+u32 vlv_dsi_port_enable(struct vlv_dsi_port *port, u32 port_bits)
+{
+	u32 val;
+	if (vlv_dsi_port_is_vid_mode(port)) {
+		vlv_dsi_port_wait_for_fifo_empty(port);
+
+		/* assert ip_tg_enable signal */
+		val = REG_READ(port->offset) &
+				~LANE_CONFIGURATION_MASK;
+		val |= port_bits;
+		REG_WRITE(port->offset, val | DPI_ENABLE);
+		REG_POSTING_READ(port->offset);
+	}
+	return 0;
+}
+
+u32 vlv_dsi_port_disable(struct vlv_dsi_port *port,
+		struct dsi_config *config)
+{
+	u32 val, clk_div = 0;
+	vlv_dsi_port_wait_for_fifo_empty(port);
+
+	if (config != NULL)
+		clk_div = config->ctx.escape_clk_div;
+
+	if (vlv_dsi_port_is_vid_mode(port)) {
+		/* de-assert ip_tg_enable signal */
+		val = REG_READ(port->offset);
+		REG_WRITE(port->offset, val & ~DPI_ENABLE);
+		REG_POSTING_READ(port->offset);
+
+		usleep_range(2000, 2500);
+	}
+
+	/* Panel commands can be sent when clock is in LP11 */
+	REG_WRITE(port->dr_offset, 0x0);
+
+	val = REG_READ(port->offset);
+	val &= ~ESCAPE_CLOCK_DIVIDER_MASK;
+	REG_WRITE(port->offset, val |
+		clk_div <<
+		ESCAPE_CLOCK_DIVIDER_SHIFT);
+
+	val = REG_READ(port->func_prg_offset);
+	val &= ~VID_MODE_FORMAT_MASK;
+	REG_WRITE(port->func_prg_offset, val);
+
+	REG_WRITE(port->eot_offset, CLOCKSTOP);
+
+	REG_WRITE(port->dr_offset, 0x1);
+
+	return 0;
+}
+
+bool vlv_dsi_port_can_be_disabled(struct vlv_dsi_port *port)
+{
+	u32 val = REG_READ(port->offset);
+	u32 func = REG_READ(port->func_prg_offset);
+	bool ret = false;
+
+	if ((val & DPI_ENABLE) || (func & CMD_MODE_DATA_WIDTH_MASK)) {
+		if (REG_READ(port->dr_offset) & DEVICE_READY)
+			ret = true;
+	}
+
+	return ret;
+}
+
+bool vlv_dsi_port_init(struct vlv_dsi_port *port, enum port enum_port,
+		enum pipe pipe)
+{
+	port->offset = MIPI_PORT_CTRL(pipe);
+	port->dr_offset = MIPI_DEVICE_READY(pipe);
+	port->func_prg_offset = MIPI_DSI_FUNC_PRG(pipe);
+	port->eot_offset = MIPI_EOT_DISABLE(pipe);
+	port->fifo_stat_offset = MIPI_GEN_FIFO_STAT(pipe);
+	port->ctrl_offset = MIPI_CTRL(pipe);
+	port->dphy_param_offset = MIPI_DPHY_PARAM(pipe);
+	port->dpi_res_offset = MIPI_DPI_RESOLUTION(pipe);
+	port->hfp_count_offset = MIPI_HFP_COUNT(pipe);
+	port->hsync_count_offset = MIPI_HSYNC_PADDING_COUNT(pipe);
+	port->hbp_count_offset = MIPI_HBP_COUNT(pipe);
+	port->vfp_count_offset = MIPI_VFP_COUNT(pipe);
+	port->vsync_count_offset = MIPI_VSYNC_PADDING_COUNT(pipe);
+	port->vbp_count_offset = MIPI_VBP_COUNT(pipe);
+	port->max_ret_pkt_offset	= MIPI_MAX_RETURN_PKT_SIZE(pipe);
+	port->hs_tx_timeout_offset	= MIPI_HS_TX_TIMEOUT(pipe);
+	port->lp_rx_timeout_offset	= MIPI_LP_RX_TIMEOUT(pipe);
+	port->ta_timeout_offset		= MIPI_TURN_AROUND_TIMEOUT(pipe);
+	port->device_reset_timer_offset	= MIPI_DEVICE_RESET_TIMER(pipe);
+	port->init_count_offset		= MIPI_INIT_COUNT(pipe);
+	port->hl_switch_count_offset	= MIPI_HIGH_LOW_SWITCH_COUNT(pipe);
+	port->lp_byteclk_offset		= MIPI_LP_BYTECLK(pipe);
+	port->dbi_bw_offset		= MIPI_DBI_BW_CTRL(pipe);
+	port->lane_switch_time_offset	= MIPI_CLK_LANE_SWITCH_TIME_CNT(pipe);
+	port->video_mode_offset		= MIPI_VIDEO_MODE_FORMAT(pipe);
+
+	/* cmd mode offsets */
+	port->hs_ls_dbi_enable_offset = MIPI_HS_LP_DBI_ENABLE(pipe);
+	port->hs_gen_ctrl_offset = MIPI_HS_GEN_CTRL(pipe);
+	port->lp_gen_ctrl_offset = MIPI_LP_GEN_CTRL(pipe);
+	port->hs_gen_data_offset = MIPI_HS_GEN_DATA(pipe);
+	port->lp_gen_data_offset = MIPI_LP_GEN_DATA(pipe);
+	port->dpi_ctrl_offset    = MIPI_DPI_CONTROL(pipe);
+
+	port->intr_stat_offset = MIPI_INTR_STAT(pipe);
+
+	port->port_id = enum_port;
+
+	/* FIXME: take appropriate value based on detection */
+	port->operation_mode = 0;
+
+	return true;
+}
+
+bool vlv_dsi_port_destroy(struct vlv_dsi_port *port)
+{
+	return true;
+}
diff --git a/drivers/video/adf/intel/core/vlv/vlv_dsi_port_cmd.c b/drivers/video/adf/intel/core/vlv/vlv_dsi_port_cmd.c
new file mode 100644
index 0000000..347428d
--- /dev/null
+++ b/drivers/video/adf/intel/core/vlv/vlv_dsi_port_cmd.c
@@ -0,0 +1,386 @@
+/*
+ * Copyright © 2013 Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Author: Jani Nikula <jani.nikula@intel.com>
+ */
+/* FIXME: remove unnecessary */
+#include <linux/export.h>
+#include <intel_adf_device.h>
+#include <video/mipi_display.h>
+#include <core/common/dsi/dsi_config.h>
+#include <core/common/dsi/dsi_pipe.h>
+#include <core/vlv/vlv_dc_regs.h>
+
+#include <drm/drmP.h>
+#include <core/common/dsi/dsi_pipe.h>
+#include <core/vlv/vlv_dc_config.h>
+#include <core/vlv/vlv_dsi_port.h>
+
+/*
+ * XXX: MIPI_DATA_ADDRESS, MIPI_DATA_LENGTH, MIPI_COMMAND_LENGTH, and
+ * MIPI_COMMAND_ADDRESS registers.
+ *
+ * Apparently these registers provide a MIPI adapter level way to send (lots of)
+ * commands and data to the receiver, without having to write the commands and
+ * data to MIPI_{HS,LP}_GEN_{CTRL,DATA} registers word by word.
+ *
+ * Presumably for anything other than MIPI_DCS_WRITE_MEMORY_START and
+ * MIPI_DCS_WRITE_MEMORY_CONTINUE (which are used to update the external
+ * framebuffer in command mode displays) these are just an optimization that can
+ * come later.
+ *
+ * For memory writes, these should probably be used for performance.
+ */
+
+static void print_stat(struct vlv_dsi_port *port)
+{
+	u32 val;
+
+	val = REG_READ(port->intr_stat_offset);
+		/* FIXME: hardcoded to pipe A for now */
+#define STAT_BIT(val, bit) (val) & (bit) ? " " #bit : ""
+	pr_info("MIPI_INTR_STAT(0) = %08x\n"
+	"%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s\n",
+		val,
+		STAT_BIT(val, TEARING_EFFECT),
+		STAT_BIT(val, SPL_PKT_SENT_INTERRUPT),
+		STAT_BIT(val, GEN_READ_DATA_AVAIL),
+		STAT_BIT(val, LP_GENERIC_WR_FIFO_FULL),
+		STAT_BIT(val, HS_GENERIC_WR_FIFO_FULL),
+		STAT_BIT(val, RX_PROT_VIOLATION),
+		STAT_BIT(val, RX_INVALID_TX_LENGTH),
+		STAT_BIT(val, ACK_WITH_NO_ERROR),
+		STAT_BIT(val, TURN_AROUND_ACK_TIMEOUT),
+		STAT_BIT(val, LP_RX_TIMEOUT),
+		STAT_BIT(val, HS_TX_TIMEOUT),
+		STAT_BIT(val, DPI_FIFO_UNDERRUN),
+		STAT_BIT(val, LOW_CONTENTION),
+		STAT_BIT(val, HIGH_CONTENTION),
+		STAT_BIT(val, TXDSI_VC_ID_INVALID),
+		STAT_BIT(val, TXDSI_DATA_TYPE_NOT_RECOGNISED),
+		STAT_BIT(val, TXCHECKSUM_ERROR),
+		STAT_BIT(val, TXECC_MULTIBIT_ERROR),
+		STAT_BIT(val, TXECC_SINGLE_BIT_ERROR),
+		STAT_BIT(val, TXFALSE_CONTROL_ERROR),
+		STAT_BIT(val, RXDSI_VC_ID_INVALID),
+		STAT_BIT(val, RXDSI_DATA_TYPE_NOT_REGOGNISED),
+		STAT_BIT(val, RXCHECKSUM_ERROR),
+		STAT_BIT(val, RXECC_MULTIBIT_ERROR),
+		STAT_BIT(val, RXECC_SINGLE_BIT_ERROR),
+		STAT_BIT(val, RXFALSE_CONTROL_ERROR),
+		STAT_BIT(val, RXHS_RECEIVE_TIMEOUT_ERROR),
+		STAT_BIT(val, RX_LP_TX_SYNC_ERROR),
+		STAT_BIT(val, RXEXCAPE_MODE_ENTRY_ERROR),
+		STAT_BIT(val, RXEOT_SYNC_ERROR),
+		STAT_BIT(val, RXSOT_SYNC_ERROR),
+		STAT_BIT(val, RXSOT_ERROR));
+#undef STAT_BIT
+}
+
+enum dsi_type {
+	DSI_DCS,
+	DSI_GENERIC,
+};
+
+/* enable or disable command mode hs transmissions */
+void vlv_dsi_port_cmd_hs_mode_enable(struct vlv_dsi_port *port, bool enable)
+{
+	u32 temp;
+	u32 mask = DBI_FIFO_EMPTY;
+
+	if (wait_for((REG_READ(port->fifo_stat_offset) & mask) == mask, 50))
+		pr_err("Timeout waiting for DBI FIFO empty\n");
+
+	temp = REG_READ(port->hs_ls_dbi_enable_offset);
+	temp &= DBI_HS_LP_MODE_MASK;
+	REG_WRITE(port->hs_ls_dbi_enable_offset,
+		enable ? DBI_HS_MODE : DBI_LP_MODE);
+
+	port->hs_mode = enable;
+}
+
+static int vlv_dsi_port_cmd_vc_send_short(struct vlv_dsi_port *port,
+		int channel, u8 data_type, u16 data)
+{
+	u32 ctrl_reg;
+	u32 ctrl;
+	u32 mask;
+
+	pr_info("channel %d, data_type %u, data %04x\n",
+			channel, data_type, data);
+
+	if (port->hs_mode) {
+		ctrl_reg = port->hs_gen_ctrl_offset;
+		mask = HS_CTRL_FIFO_FULL;
+	} else {
+		ctrl_reg = port->lp_gen_ctrl_offset;
+		mask = LP_CTRL_FIFO_FULL;
+	}
+
+	if (wait_for((REG_READ(port->fifo_stat_offset) & mask) == 0, 50)) {
+		pr_err("Timeout waiting for HS/LP CTRL FIFO !full\n");
+		print_stat(port);
+	}
+
+	/*
+	 * Note: This function is also used for long packets, with length passed
+	 * as data, since SHORT_PACKET_PARAM_SHIFT ==
+	 * LONG_PACKET_WORD_COUNT_SHIFT.
+	 */
+	ctrl = data << SHORT_PACKET_PARAM_SHIFT |
+		channel << VIRTUAL_CHANNEL_SHIFT |
+		data_type << DATA_TYPE_SHIFT;
+
+	REG_WRITE(ctrl_reg, ctrl);
+
+	return 0;
+}
+
+static int vlv_dsi_port_cmd_vc_send_long(struct vlv_dsi_port *port,
+			int channel, u8 data_type, const u8 *data, int len)
+{
+	u32 data_reg;
+	int i, j, n;
+	u32 mask;
+
+	pr_info("channel %d, data_type %u, len %04x\n",
+			channel, data_type, len);
+
+	if (port->hs_mode) {
+		data_reg = port->hs_gen_data_offset;
+		mask = HS_DATA_FIFO_FULL;
+	} else {
+		data_reg = port->lp_gen_data_offset;
+		mask = LP_DATA_FIFO_FULL;
+	}
+
+	if (wait_for((REG_READ(port->fifo_stat_offset) & mask) == 0, 50))
+		pr_err("Timeout waiting for HS/LP DATA FIFO !full\n");
+
+	for (i = 0; i < len; i += n) {
+		u32 val = 0;
+		n = min_t(int, len - i, 4);
+
+		for (j = 0; j < n; j++)
+			val |= *data++ << 8 * j;
+
+		REG_WRITE(data_reg, val);
+		/*
+		 * XXX: check for data fifo full, once that is set, write 4
+		 * dwords, then wait for not set, then continue.
+		 */
+	}
+
+	return vlv_dsi_port_cmd_vc_send_short(port, channel, data_type, len);
+}
+
+static int vlv_dsi_port_cmd_vc_write_common(struct vlv_dsi_port *port,
+		int channel, const u8 *data, int len, enum dsi_type type)
+{
+	int ret;
+
+	if (len == 0) {
+		BUG_ON(type == DSI_GENERIC);
+		ret = vlv_dsi_port_cmd_vc_send_short(port, channel,
+				MIPI_DSI_GENERIC_SHORT_WRITE_0_PARAM, 0);
+	} else if (len == 1) {
+		ret = vlv_dsi_port_cmd_vc_send_short(port, channel,
+					type == DSI_GENERIC ?
+					MIPI_DSI_GENERIC_SHORT_WRITE_1_PARAM :
+					MIPI_DSI_DCS_SHORT_WRITE, data[0]);
+	} else if (len == 2) {
+		ret = vlv_dsi_port_cmd_vc_send_short(port, channel,
+					type == DSI_GENERIC ?
+					MIPI_DSI_GENERIC_SHORT_WRITE_2_PARAM :
+					MIPI_DSI_DCS_SHORT_WRITE_PARAM,
+					(data[1] << 8) | data[0]);
+	} else {
+		ret = vlv_dsi_port_cmd_vc_send_long(port, channel,
+					type == DSI_GENERIC ?
+					MIPI_DSI_GENERIC_LONG_WRITE :
+					MIPI_DSI_DCS_LONG_WRITE, data, len);
+	}
+
+	return ret;
+}
+
+int vlv_dsi_port_cmd_vc_dcs_write(struct vlv_dsi_port *port, int channel,
+		const u8 *data, int len)
+{
+	return vlv_dsi_port_cmd_vc_write_common(port, channel, data, len,
+		DSI_DCS);
+}
+
+int vlv_dsi_port_cmd_vc_generic_write(struct vlv_dsi_port *port, int channel,
+		const u8 *data, int len)
+{
+	return vlv_dsi_port_cmd_vc_write_common(port, channel, data, len,
+		DSI_GENERIC);
+}
+
+static int vlv_dsi_port_cmd_vc_dcs_send_read_request(struct vlv_dsi_port *port,
+		int channel, u8 dcs_cmd)
+{
+	return vlv_dsi_port_cmd_vc_send_short(port, channel, MIPI_DSI_DCS_READ,
+				dcs_cmd);
+}
+
+static int vlv_dsi_port_cmd_vc_generic_send_read_request(
+		struct vlv_dsi_port *port,
+		int channel, u8 *reqdata, int reqlen)
+{
+	u16 data;
+	u8 data_type;
+
+	switch (reqlen) {
+	case 0:
+		data_type = MIPI_DSI_GENERIC_READ_REQUEST_0_PARAM;
+		data = 0;
+		break;
+	case 1:
+		data_type = MIPI_DSI_GENERIC_READ_REQUEST_1_PARAM;
+		data = reqdata[0];
+		break;
+	case 2:
+		data_type = MIPI_DSI_GENERIC_READ_REQUEST_2_PARAM;
+		data = (reqdata[1] << 8) | reqdata[0];
+		break;
+	default:
+		BUG();
+	}
+
+	return vlv_dsi_port_cmd_vc_send_short(port, channel, data_type, data);
+}
+
+static int vlv_dsi_port_cmd_read_data_return(struct vlv_dsi_port *port,
+		u8 *buf, int buflen)
+{
+	int i, len = 0;
+	u32 data_reg, val;
+
+	if (port->hs_mode)
+		data_reg = port->hs_gen_data_offset;
+	else
+		data_reg = port->lp_gen_data_offset;
+
+	while (len < buflen) {
+		val = REG_READ(data_reg);
+		for (i = 0; i < 4 && len < buflen; i++, len++)
+			buf[len] = val >> 8 * i;
+	}
+
+	return len;
+}
+
+int vlv_dsi_port_cmd_vc_dcs_read(struct vlv_dsi_port *port, int channel,
+		u8 dcs_cmd, u8 *buf, int buflen)
+{
+	u32 mask;
+	int ret;
+
+	/*
+	 * XXX: should issue multiple read requests and reads if request is
+	 * longer than MIPI_MAX_RETURN_PKT_SIZE
+	 */
+
+	REG_WRITE(port->intr_stat_offset, GEN_READ_DATA_AVAIL);
+
+	ret = vlv_dsi_port_cmd_vc_dcs_send_read_request(port, channel, dcs_cmd);
+	if (ret)
+		return ret;
+
+	mask = GEN_READ_DATA_AVAIL;
+	if (wait_for((REG_READ(port->intr_stat_offset) & mask) == mask, 50))
+		DRM_ERROR("Timeout waiting for read data.\n");
+
+	ret = vlv_dsi_port_cmd_read_data_return(port, buf, buflen);
+	if (ret < 0)
+		return ret;
+
+	if (ret != buflen)
+		return -EIO;
+
+	return 0;
+}
+
+int vlv_dsi_port_cmd_vc_generic_read(struct vlv_dsi_port *port, int channel,
+		u8 *reqdata, int reqlen, u8 *buf, int buflen)
+{
+	u32 mask;
+	int ret;
+
+	/*
+	 * XXX: should issue multiple read requests and reads if request is
+	 * longer than MIPI_MAX_RETURN_PKT_SIZE
+	 */
+
+	REG_WRITE(port->intr_stat_offset, GEN_READ_DATA_AVAIL);
+
+	ret = vlv_dsi_port_cmd_vc_generic_send_read_request(port, channel,
+				reqdata, reqlen);
+	if (ret)
+		return ret;
+
+	mask = GEN_READ_DATA_AVAIL;
+	if (wait_for((REG_READ(port->intr_stat_offset) & mask) == mask, 50))
+		pr_err("Timeout waiting for read data.\n");
+
+	ret = vlv_dsi_port_cmd_read_data_return(port, buf, buflen);
+	if (ret < 0)
+		return ret;
+
+	if (ret != buflen)
+		return -EIO;
+
+	return 0;
+}
+
+/*
+ * send a video mode command
+ *
+ * XXX: commands with data in MIPI_DPI_DATA?
+ */
+int vlv_dsi_port_cmd_dpi_send_cmd(struct vlv_dsi_port *port, u32 cmd, bool hs)
+{
+	u32 mask;
+
+	/* XXX: pipe, hs */
+	if (hs)
+		cmd &= ~DPI_LP_MODE;
+	else
+		cmd |= DPI_LP_MODE;
+
+	/* clear bit */
+	REG_WRITE(port->intr_stat_offset, SPL_PKT_SENT_INTERRUPT);
+
+	/* XXX: old code skips write if control unchanged */
+	if (cmd == REG_READ(port->dpi_ctrl_offset))
+		pr_err("Same special packet %02x twice in a row.\n", cmd);
+
+	REG_WRITE(port->dpi_ctrl_offset, cmd);
+
+	mask = SPL_PKT_SENT_INTERRUPT;
+	if (wait_for((REG_READ(port->intr_stat_offset) & mask) == mask, 100))
+		pr_err("Video mode command 0x%08x send failed.\n", cmd);
+
+	return 0;
+}
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dc_config.h b/drivers/video/adf/intel/include/core/vlv/vlv_dc_config.h
index dfa32c6..db8bb6d 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_dc_config.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dc_config.h
@@ -24,6 +24,7 @@
 #include <core/vlv/vlv_pri_plane.h>
 #include <core/vlv/vlv_sp_plane.h>
 #include <core/vlv/vlv_dpst.h>
+#include <core/vlv/vlv_dsi_port.h>
 
 #define VLV_N_PLANES	6
 #define VLV_N_PIPES	2
@@ -132,6 +133,10 @@ extern int vlv_display_off(struct intel_pipe *pipe);
 int vlv_num_planes_enabled(struct intel_pipe *pipe);
 void vlv_update_plane_status(struct intel_pipe *pipe, int plane, bool enabled);
 
+/* port export functions */
+bool vlv_dsi_port_init(struct vlv_dsi_port *port, enum port, enum pipe);
+bool vlv_dsi_port_destroy(struct vlv_dsi_port *port);
+
 /* vlv_debugfs */
 extern int vlv_debugfs_init(struct vlv_dc_config *vlv_config);
 extern void vlv_debugfs_teardown(struct vlv_dc_config *vlv_config);
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dsi_port.h b/drivers/video/adf/intel/include/core/vlv/vlv_dsi_port.h
new file mode 100644
index 0000000..56dea3e
--- /dev/null
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dsi_port.h
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2014, Intel Corporation.
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _VLV_DSI_PORT_H_
+#define _VLV_DSI_PORT_H_
+
+#include <core/common/dsi/dsi_pipe.h>
+#include <core/vlv/vlv_dc_config.h>
+
+struct vlv_dsi_port {
+	enum port port_id;
+	u32 offset;
+	u32 operation_mode;
+	u32 dr_offset; /* device_ready */
+	u32 eot_offset;
+	u32 func_prg_offset;
+	u32 fifo_stat_offset;
+	u32 ctrl_offset;
+	u32 dphy_param_offset;
+	u32 dpi_res_offset;
+	u32 hactive_count_offset;
+	u32 hfp_count_offset;
+	u32 hsync_count_offset;
+	u32 hbp_count_offset;
+	u32 vfp_count_offset;
+	u32 vsync_count_offset;
+	u32 vbp_count_offset;
+	u32 max_ret_pkt_offset;
+	u32 hs_tx_timeout_offset;
+	u32 lp_rx_timeout_offset;
+	u32 ta_timeout_offset;
+	u32 device_reset_timer_offset;
+	u32 init_count_offset;
+	u32 hl_switch_count_offset;
+	u32 lp_byteclk_offset;
+	u32 dbi_bw_offset;
+	u32 lane_switch_time_offset;
+	u32 video_mode_offset;
+
+	/* cmd mode offsets */
+	bool hs_mode;
+	u32 hs_ls_dbi_enable_offset;
+	u32 hs_gen_ctrl_offset;
+	u32 lp_gen_ctrl_offset;
+	u32 hs_gen_data_offset;
+	u32 lp_gen_data_offset;
+	u32 dpi_ctrl_offset;
+
+	u32 intr_stat_offset;/*FIXME: to be moved to interrupt obj */
+};
+
+u32  vlv_dsi_port_prepare(struct vlv_dsi_port *port,
+		struct drm_mode_modeinfo *mode,
+		struct dsi_config *config);
+u32 vlv_dsi_port_pre_enable(struct vlv_dsi_port *port,
+		struct drm_mode_modeinfo *mode,
+		struct dsi_config *config);
+u32 vlv_dsi_port_enable(struct vlv_dsi_port *port, u32 port_bits);
+u32 vlv_dsi_port_disable(struct vlv_dsi_port *port,
+		struct dsi_config *config);
+bool vlv_dsi_port_set_device_ready(struct vlv_dsi_port *port);
+bool vlv_dsi_port_clear_device_ready(struct vlv_dsi_port *port);
+u32  vlv_dsi_port_wait_for_fifo_empty(struct vlv_dsi_port *port);
+bool vlv_dsi_port_can_be_disabled(struct vlv_dsi_port *port);
+u32 vlv_dsi_port_is_vid_mode(struct vlv_dsi_port *port);
+u32 vlv_dsi_port_is_cmd_mode(struct vlv_dsi_port *port);
+void vlv_dsi_port_cmd_hs_mode_enable(struct vlv_dsi_port *port, bool enable);
+int vlv_dsi_port_cmd_vc_dcs_write(struct vlv_dsi_port *port, int channel,
+		const u8 *data, int len);
+int vlv_dsi_port_cmd_vc_generic_write(struct vlv_dsi_port *port, int channel,
+		const u8 *data, int len);
+int vlv_dsi_port_cmd_vc_dcs_read(struct vlv_dsi_port *port, int channel,
+		u8 dcs_cmd, u8 *buf, int buflen);
+int vlv_dsi_port_cmd_vc_generic_read(struct vlv_dsi_port *port, int channel,
+		u8 *reqdata, int reqlen, u8 *buf, int buflen);
+int vlv_dsi_port_cmd_dpi_send_cmd(struct vlv_dsi_port *port, u32 cmd, bool hs);
+void vlv_dsi_port_cmd_hs_mode_enable(struct vlv_dsi_port *port,
+			bool enable);
+int vlv_dsi_port_cmd_vc_dcs_write(struct vlv_dsi_port *port, int channel,
+			const u8 *data, int len);
+int vlv_dsi_port_cmd_vc_generic_write(struct vlv_dsi_port *port, int channel,
+			const u8 *data, int len);
+int vlv_dsi_port_cmd_vc_dcs_read(struct vlv_dsi_port *port, int channel,
+			u8 dcs_cmd, u8 *buf, int buflen);
+int vlv_dsi_port_cmd_vc_generic_read(struct vlv_dsi_port *port, int channel,
+			u8 *reqdata, int reqlen, u8 *buf, int buflen);
+int vlv_dsi_port_cmd_dpi_send_cmd(struct vlv_dsi_port *port, u32 cmd, bool hs);
+
+
+#endif /* _VLV_DSI_PORT_H_ */
-- 
1.7.9.5

